File: bento/src/main/java/com/yelp/android/bento/components/TabViewPagerComponent.java
package com.yelp.android.bento.components;

import androidx.annotation.ColorRes;
import androidx.annotation.DimenRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;
import com.yelp.android.bento.componentcontrollers.TabViewPagerComponentController;
import com.yelp.android.bento.componentcontrollers.TabViewPagerComponentViewHolder;
import com.yelp.android.bento.componentcontrollers.TabViewPagerComponentViewHolder.TabViewPagerComponentViewHolderData;
import com.yelp.android.bento.core.Component;
import com.yelp.android.bento.core.ComponentViewHolder;
import java.util.List;

public class TabViewPagerComponent extends Component implements ViewPager.OnPageChangeListener {

    private final TabViewPagerComponentViewModel mViewModel;
    private final TabViewPagerComponentController mViewPagerComponentController;
    private final TabViewPagerComponentViewHolderData mTabViewPagerComponentViewHolderData;
    private final TabViewPagerOnPageChangeListener mOnPageChangeListener;

    public TabViewPagerComponent(
            @NonNull TabViewPagerComponentViewModel viewModel,
            @Nullable TabViewPagerOnPageChangeListener onPageChangeListener) {
        mViewModel = viewModel;
        mOnPageChangeListener = onPageChangeListener;
        mViewPagerComponentController = new TabViewPagerComponentController();
        mViewPagerComponentController.setPageTitles(mViewModel.getPageTitles());
        mViewPagerComponentController.addAll(mViewModel.getComponentList());
        mTabViewPagerComponentViewHolderData =
                new TabViewPagerComponentViewHolderData(mViewModel, mViewPagerComponentController);
    }

    @Override
    public ViewPager.OnPageChangeListener getPresenter(int position) {
        return this;
    }

    @Override
    public TabViewPagerComponentViewHolderData getItem(int position) {
        return mTabViewPagerComponentViewHolderData;
    }

    @Override
    public int getCount() {
        return 1;
    }

    @Override
    public Class<? extends ComponentViewHolder> getHolderType(int position) {
        return TabViewPagerComponentViewHolder.class;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {}

    @Override
    public void onPageSelected(int position) {
        mViewModel.setCurrentIndex(position);
        if (mOnPageChangeListener != null) {
            mOnPageChangeListener.onPageSelected(position);
        }
    }

    @Override
    public void onPageScrollStateChanged(int state) {}

    public static class TabViewPagerComponentViewModel {
        private List<String> mPageTitles;
        private List<Component> mComponentList;
        private int mCurrentIndex;

        @DimenRes private int mTabLayoutSideMargins;
        @DrawableRes private int mBackground;
        @DrawableRes private int mTabLayoutBackground;
        @ColorRes private int mTabLayoutTextColorNormal;
        @ColorRes private int mTabLayoutTextColorSelected;
        @ColorRes private int mTabLayoutIndicatorColor;

        public TabViewPagerComponentViewModel(
                @NonNull List<String> pageTitles, @NonNull List<Component> componentsList) {
            if (pageTitles.size() == componentsList.size()) {
                mPageTitles = pageTitles;
                mComponentList = componentsList;
            } else {
                throw new IllegalArgumentException(
                        "Number of titles must match number of components");
            }
        }

        public List<String> getPageTitles() {
            return mPageTitles;
        }

        public List<Component> getComponentList() {
            return mComponentList;
        }

        public int getCurrentIndex() {
            return mCurrentIndex;
        }

        public void setCurrentIndex(int currentIndex) {
            this.mCurrentIndex = currentIndex;
        }

        public int getTabLayoutSideMargins() {
            return mTabLayoutSideMargins;
        }

        public TabViewPagerComponentViewModel setTabLayoutSideMargins(@DimenRes int sideMargins) {
            mTabLayoutSideMargins = sideMargins;
            return this;
        }

        @DrawableRes
        public int getBackground() {
            return mBackground;
        }

        public TabViewPagerComponentViewModel setBackground(@DrawableRes int background) {
            mBackground = background;
            return this;
        }

        @DrawableRes
        public int getTabLayoutBackground() {
            return mTabLayoutBackground;
        }

        public TabViewPagerComponentViewModel setTabLayoutBackground(@DrawableRes int background) {
            mTabLayoutBackground = background;
            return this;
        }

        @ColorRes
        public int getTextColorNormal() {
            return mTabLayoutTextColorNormal;
        }

        public TabViewPagerComponentViewModel setTextColorNormal(
                @ColorRes int tabLayoutTextNormal) {
            mTabLayoutTextColorNormal = tabLayoutTextNormal;
            return this;
        }

        @ColorRes
        public int getTextColorSelected() {
            return mTabLayoutTextColorSelected;
        }

        public TabViewPagerComponentViewModel setTextColorSelected(
                @ColorRes int tabLayoutTextSelected) {
            mTabLayoutTextColorSelected = tabLayoutTextSelected;
            return this;
        }

        @ColorRes
        public int getIndicatorColor() {
            return mTabLayoutIndicatorColor;
        }

        public TabViewPagerComponentViewModel setIndicatorColor(
                @ColorRes int tabLayoutIndicatorColor) {
            mTabLayoutIndicatorColor = tabLayoutIndicatorColor;
            return this;
        }
    }

    public interface TabViewPagerOnPageChangeListener {

        void onPageSelected(int position);
    }
}

File: bento/src/main/java/com/yelp/android/bento/components/TabViewPagerComponent.kt
package com.yelp.android.bento.components

import androidx.annotation.ColorRes
import androidx.annotation.DimenRes
import androidx.annotation.DrawableRes
import androidx.viewpager.widget.ViewPager
import com.yelp.android.bento.componentcontrollers.TabViewPagerComponentController
import com.yelp.android.bento.componentcontrollers.TabViewPagerComponentViewHolder
import com.yelp.android.bento.componentcontrollers.TabViewPagerComponentViewHolder.TabViewPagerComponentViewHolderData
import com.yelp.android.bento.core.Component
import com.yelp.android.bento.core.ComponentViewHolder

class TabViewPagerComponent(
    private val viewModel: TabViewPagerComponentViewModel,
    private val onPageChangeListener: TabViewPagerOnPageChangeListener?
) : Component(), ViewPager.OnPageChangeListener {
    private val viewPagerComponentController: TabViewPagerComponentController =
            TabViewPagerComponentController()
    private val tabViewPagerComponentViewHolderData: TabViewPagerComponentViewHolderData

    init {
        viewPagerComponentController.pageTitles = viewModel.pageTitles
        viewPagerComponentController.addAll(viewModel.componentList)
        tabViewPagerComponentViewHolderData =
                TabViewPagerComponentViewHolderData(viewModel, viewPagerComponentController)
    }

    override fun getPresenter(position: Int): ViewPager.OnPageChangeListener? = this

    override fun getItem(position: Int) = tabViewPagerComponentViewHolderData

    override fun getCount() = 1

    override fun getHolderType(position: Int): Class<out ComponentViewHolder<*, *>> =
        TabViewPagerComponentViewHolder::class.java

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

    override fun onPageSelected(position: Int) {
        viewModel.currentIndex = position
        onPageChangeListener?.onPageSelected(position)
    }

    override fun onPageScrollStateChanged(state: Int) {}

    data class TabViewPagerComponentViewModel(
        val pageTitles: List<String>,
        val componentsList: List<Component>,
        var currentIndex: Int = 0,
        @DimenRes var tabLayoutSideMargins: Int = 0,
        @DrawableRes var background: Int = 0,
        @DrawableRes var tabLayoutBackground: Int = 0,
        @ColorRes var tabLayoutTextColorNormal: Int = 0,
        @ColorRes var tabLayoutTextColorSelected: Int = 0,
        @ColorRes var tabLayoutIndicatorColor: Int = 0
    ) {
        var tabTitles: List<String>
        var componentList: List<Component>

        init {
            /**
             * The count of tabTitles or pageTitles need to be the same as componentsList
             * because there must be one component per tab.
             */
            if (pageTitles.size == componentsList.size) {
                tabTitles = pageTitles
                componentList = componentsList
            } else {
                throw IllegalArgumentException(
                        "Number of titles must match number of components")
            }
        }
    }

    interface TabViewPagerOnPageChangeListener {

        fun onPageSelected(position: Int)
    }
}
